File: src/main/java/com/zero_delusions/dev_assignment/network/PokemonTradeManager.java
package com.zero_delusions.dev_assignment.network;


import com.cobblemon.mod.common.pokemon.Pokemon;
import com.google.gson.Gson;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPubSub;

import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public class PokemonTradeManager {
    private static final ConcurrentHashMap<String, TradeRequest> pendingTrades = new ConcurrentHashMap<>();
    private static final int TRADE_TIMEOUT_SECONDS = 120;

    public static class TradeRequest {
        private final String requesterId;
        private final String requesterServer;
        private final int requesterPokemonSlot;
        private final String targetId;
        private final String targetServer;
        private final long timestamp;

        public TradeRequest(
            String requesterId,
            String requesterServer,
            int requesterPokemonSlot,
            String targetId,
            String targetServer
        ) {
            this(
                requesterId,
                requesterServer,
                requesterPokemonSlot,
                targetId, targetServer,
                System.currentTimeMillis()
            );
        }

        public TradeRequest(String requesterId, String requesterServer, int requesterPokemonSlot,
        String targetId, String targetServer, long timestamp) {
            this.requesterId = requesterId;
            this.requesterServer = requesterServer;
            this.requesterPokemonSlot = requesterPokemonSlot;
            this.targetId = targetId;
            this.targetServer = targetServer;
            this.timestamp = timestamp;
        }

        // Getters
        public String getRequesterId() { return requesterId; }
        public String getRequesterServer() { return requesterServer; }
        public int getRequesterPokemonSlot() { return requesterPokemonSlot; }
        public String getTargetId() { return targetId; }
        public String getTargetServer() { return targetServer; }
        public long getTimestamp() { return timestamp; }
    }

    public static boolean requestTrade(
        String requesterId,
        String requesterServer,
        int requesterPokemonSlot,
        String targetId,
        String targetServer
    ) {
        String tradeId = "trade:" + UUID.randomUUID();

        // Get the requester's Pokémon from the specified slot
        Pokemon pokemon = getPlayerPokemonFromSlot(targetServer, requesterId, requesterPokemonSlot);
        if (!canTradePokemon(pokemon)) return false;

        // Lock the Pokémon to prevent other trades
        lockPokemon(targetServer, requesterId, requesterPokemonSlot);

        TradeRequest request = new TradeRequest(
            requesterId,
            requesterServer,
            requesterPokemonSlot,
            targetId,
            targetServer
        );
        pendingTrades.put(tradeId, request);

        scheduleTradeTimeout(tradeId);

        if (requesterServer.equals(targetServer)) {
            // If both players are on the same server, notify directly
            notifyPlayerOfTradeRequest(targetId, requesterId, tradeId);
        } else {
            // Otherwise, send a network request to the target server with trade details
            try (JedisPool pool = new JedisPool();
                 var redis = pool.getResource()) {

                    String dataKey = "pokemontrade:request:" + tradeId;

                    Gson gson = new Gson();
                    redis.setex(dataKey, TRADE_TIMEOUT_SECONDS, gson.toJson(request));

                    new NetworkRequest(
                            targetServer,
                    "trade_request|" + dataKey + "|" + tradeId
                    ).sendAndWait(5000);
                }
            }
            return true;
        }

    public static boolean acceptTrade(String tradeId, String targetId, int targetPokemonSlot) {
        TradeRequest request = pendingTrades.get(tradeId);
        if (request == null || !request.getTargetId().equals(targetId)) return false;

        // Get the target's Pokémon from the specified slot
        Pokemon targetPokemon = getPlayerPokemonFromSlot(
                request.getRequesterServer(), targetId, targetPokemonSlot
        );
        if (!canTradePokemon(targetPokemon)) return false;

        lockPokemon(request.getRequesterServer(), targetId, targetPokemonSlot);

        if (request.getRequesterServer().equals(request.getTargetServer())) {
            // If both players are on the same server, execute the trade immediately
            executeTrade(
                request.getRequesterId(),
                request.getRequesterPokemonSlot(),
                targetId,
                targetPokemonSlot
            );
            notifyTradeComplete(request.getRequesterId(), targetId);
        } else {
            // For cross-server trades, send a network request
            try (JedisPool pool = new JedisPool();
                var redis = pool.getResource()) {

                    String targetPokemonData = serializePokemon(targetPokemon);
                    String tradeDataKey = "pokemontrade:data:" + tradeId;
                    redis.setex(tradeDataKey, 60, targetPokemonData);

                    String response = new NetworkRequest(
                        request.getRequesterServer(),
                        "trade_accept|" + tradeDataKey + "|" + tradeId
                    ).sendAndWait(10000);

                    if ("COMPLETE".equals(response)) {
                        // On trade complete replace Pokémon
                        removePokemonFromPlayer(targetId, targetPokemonSlot);

                        String requesterPokemonKey = response.split("\\|")[1];
                        Pokemon requesterPokemon = deserializePokemon(redis.get(requesterPokemonKey));

                        addPokemonToPlayer(targetId, requesterPokemon);
                        notifyTradeComplete(targetId);
                    } else {
                        // If the trade was not successful, unlock the Pokémon and notify failure
                        unlockPokemon(targetId, targetPokemonSlot);
                        notifyTradeFailed(targetId);
                    }
                }
            }

            pendingTrades.remove(tradeId);
            return true;
        }

    public static boolean denyTrade(String tradeId, String targetId) {
        TradeRequest request = pendingTrades.get(tradeId);
        if (request == null || !request.getTargetId().equals(targetId)) return false;

        if (request.getRequesterServer().equals(request.getTargetServer())) {
            // For same-server trades, unlock the Pokémon reserved by the requester
            unlockPokemon(request.getRequesterId(), request.getRequesterPokemonSlot());
        } else {
            // For cross-server trades, notify the requester server of the denial
            new NetworkRequest(
                    request.getRequesterServer(),
            "trade_deny|" + tradeId
            ).sendAndWait(10000);
        }

        pendingTrades.remove(tradeId);
        notifyTradeDenied(request.getRequesterId(), targetId);
        return true;
    }

    public static void setupTradeRequestListener(String currentServerId) {
        new Thread(() -> {
            try (JedisPool pool = new JedisPool();
                var jedis = pool.getResource()) {

                    jedis.subscribe(new JedisPubSub() {
                        @Override
                        public void onMessage(String channel, String message) {
                            if (channel.equals("server:" + currentServerId + ":requests")) {
                                String[] parts = message.split("\\|");
                                switch (parts[0]) {
                                    case "trade_request":
                                        handleTradeRequest(parts, jedis);
                                        break;
                                    case "trade_accept":
                                        handleTradeAccept(parts, jedis);
                                        break;
                                    case "trade_deny":
                                        handleTradeDeny(parts);
                                        break;
                                }
                            }
                        }
                    }, "server:" + currentServerId + ":requests");
                }
            }).start();
    }

    private static void handleTradeRequest(String[] parts, Jedis jedis) {
        String dataKey = parts[1];
        String tradeId = parts[2];

        Gson gson = new Gson();
        TradeRequest requestData = gson.fromJson(jedis.get(dataKey), TradeRequest.class);

        pendingTrades.put(tradeId, requestData);
        notifyPlayerOfTradeRequest(requestData.getTargetId(), requestData.getRequesterId(), tradeId);
    }

    private static void handleTradeAccept(String[] parts, Jedis jedis) {
        // parts[0] = "trade_accept", parts[1] = dataKey, parts[2] = tradeId, parts[last] = responseChannel
        String dataKey = parts[1];
        String tradeId = parts[2];
        TradeRequest request = pendingTrades.get(tradeId);
        if (request == null) return;

        // Get requester's Pokémon
        Pokemon requesterPokemon = getPlayerPokemonFromSlot(
                request.getRequesterServer(),
                request.getRequesterId(),
                request.getRequesterPokemonSlot()
        );

        // Get target's Pokémon from Redis
        String targetPokemonData = jedis.get(dataKey);
        Pokemon targetPokemon = deserializePokemon(targetPokemonData);

        // Store requester's Pokémon in Redis for the target server
        String requesterPokemonKey = "pokemontrade:response:" + tradeId;
        jedis.setex(requesterPokemonKey, 60, serializePokemon(requesterPokemon));

        // Execute trade on this side
        removePokemonFromPlayer(request.getRequesterId(), request.getRequesterPokemonSlot());
        addPokemonToPlayer(request.getRequesterId(), targetPokemon);

        // Respond with completion
        String responseChannel = parts[parts.length - 1];
        jedis.publish(responseChannel, "COMPLETE|" + requesterPokemonKey);

        // Notify player
        notifyTradeComplete(request.getRequesterId());

        // Remove pending trade
        pendingTrades.remove(tradeId);
    }

    private static void handleTradeDeny(String[] parts) {
        // parts[0] = "trade_deny", parts[1] = tradeId
        String tradeId = parts[1];
        TradeRequest request = pendingTrades.get(tradeId);
        if (request == null) return;

        unlockPokemon(request.getRequesterId(), request.getRequesterPokemonSlot());

        notifyTradeDenied(request.getRequesterId(), request.getTargetId());

        // Remove pending trade
        pendingTrades.remove(tradeId);
    }

    // Placeholder implementations
    private static Pokemon getPlayerPokemonFromSlot(String server, String playerId, int slot) {
        return new Pokemon();
    }
    private static boolean canTradePokemon(Pokemon pokemon) {
        return true;
    }

    private static void lockPokemon(String server, String playerId, int slot) {}
    private static void unlockPokemon(String playerId, int slot) {}

    private static void scheduleTradeTimeout(String tradeId) {}
    private static void executeTrade(String requesterId, int requesterSlot, String targetId, int targetSlot) {}

    private static void notifyPlayerOfTradeRequest(String targetId, String requesterId, String tradeId) {}
    private static void notifyTradeComplete(String playerId) {
        notifyTradeComplete(playerId, null);
    }
    private static void notifyTradeComplete(String playerId, String targetId) {}
    private static void notifyTradeFailed(String playerId) {}
    private static void notifyTradeDenied(String requesterId, String targetId) {}

    private static String serializePokemon(Pokemon pokemon) {
        return "";
    }
    private static Pokemon deserializePokemon(String data) {
        return new Pokemon();
    }

    private static void removePokemonFromPlayer(String playerId, int slot) {}
    private static void addPokemonToPlayer(String playerId, Pokemon pokemon) {}
}

File: src/main/kotlin/com/zero_delusions/dev_assignment/network/PokemonTradeManager.kt
package com.zero_delusions.dev_assignment.network

import com.cobblemon.mod.common.pokemon.Pokemon
import kotlinx.serialization.json.Json
import redis.clients.jedis.JedisPool
import redis.clients.jedis.JedisPubSub
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

object PokemonTradeManager {

    private val pendingTrades = ConcurrentHashMap<String, TradeRequest>()
    private val TRADE_TIMEOUT_SECONDS = 120

    data class TradeRequest(
        val requesterId: String,
        val requesterServer: String,
        val requesterPokemonSlot: Int,
        val targetId: String,
        val targetServer: String,
        val timestamp: Long = System.currentTimeMillis(),
    )

    fun requestTrade(
        requesterId: String,
        requesterServer: String,
        requesterPokemonSlot: Int,
        targetId: String,
        targetServer: String,
    ): Boolean {
        val tradeId = "trade:${UUID.randomUUID()}"
        // Verify the Pokémon exists and can be traded
        val pokemon = getPlayerPokemonFromSlot(targetServer, requesterId, requesterPokemonSlot) ?: return false
        if (!canTradePokemon(pokemon)) return false
        // Lock the Pokémon to prevent changes while trade is pending
        lockPokemon(targetServer, requesterId, requesterPokemonSlot)

        val request = TradeRequest(
            requesterId,
            requesterServer,
            requesterPokemonSlot,
            targetId,
            targetServer
        )
        pendingTrades[tradeId] = request

        // Set up a timeout to auto-cancel
        scheduleTradeTimeout(tradeId)

        // Send the trade request to the target server
        if (requesterServer == targetServer) {
            // Local trade - notify target player directly
            notifyPlayerOfTradeRequest(targetId, requesterId, tradeId)
        } else {
            val redis = JedisPool().resource
            val dataKey = "pokemontrade:request:$tradeId"
            redis.setex(dataKey, TRADE_TIMEOUT_SECONDS, Json.encodeToString(request))
            // Send notification to target server
            NetworkRequest(
                targetServer = targetServer,
                payload = "trade_request|$dataKey|$tradeId"
            ).sendAndWait(5000)

            redis.close()
        }

        return true
    }

    // Accepts a trade request
    fun acceptTrade(
        tradeId: String,
        targetId: String,
        targetPokemonSlot: Int,
    ): Boolean {
        val request = pendingTrades[tradeId] ?: return false

        // Verify the acceptor matches the target in the request
        if (request.targetId != targetId) return false
        // Get and validate the target Pokémon
        val targetPokemon =
            getPlayerPokemonFromSlot(request.requesterServer, targetId, targetPokemonSlot) ?: return false
        if (!canTradePokemon(targetPokemon)) return false

        // Lock the target Pokémon
        lockPokemon(request.requesterServer, targetId, targetPokemonSlot)

        // Execute the trade
        if (request.requesterServer == request.targetServer) {
            // Same server trade
            executeTrade(
                request.requesterId,
                request.requesterPokemonSlot,
                targetId,
                targetPokemonSlot
            )
            notifyTradeComplete(request.requesterId, targetId)
        } else {
            // Cross-server trade
            val redis = JedisPool().resource
            // Serialize Pokémon data
            val targetPokemonData = serializePokemon(targetPokemon)
            val tradeDataKey = "pokemontrade:data:$tradeId"
            // Store target Pokémon in Redis
            redis.setex(tradeDataKey, 60, targetPokemonData)
            // Notify requester server
            val response = NetworkRequest(
                targetServer = request.requesterServer,
                payload = "trade_accept|$tradeDataKey|$tradeId"
            ).sendAndWait(10000)

            if (response == "COMPLETE") {
                // Trade completed successfully

                // Probably instead of removing and adding Pokémon, a direct Pokémon set can be used to change Pokémon data to needed one
                removePokemonFromPlayer(targetId, targetPokemonSlot)
                // Load the requester's Pokémon from the response
                val requesterPokemonKey = response.split("|")[1]
                val requesterPokemon = deserializePokemon(redis.get(requesterPokemonKey))
                addPokemonToPlayer(targetId, requesterPokemon)

                notifyTradeComplete(targetId)
            } else {
                // Trade failed
                unlockPokemon(targetId, targetPokemonSlot)
                notifyTradeFailed(targetId)
            }

            redis.close()
        }

        // Remove pending trade
        pendingTrades.remove(tradeId)
        return true
    }

    // Denies a trade request
    fun denyTrade(tradeId: String, targetId: String): Boolean {
        val request = pendingTrades[tradeId] ?: return false

        // Verify the denier matches the target
        if (request.targetId != targetId) return false
        // Unlock the requester's Pokémon
        if (request.requesterServer == request.targetServer) {
            unlockPokemon(request.requesterId, request.requesterPokemonSlot)
        } else {
            // Notify the requester's server to unlock
            NetworkRequest(
                targetServer = request.requesterServer,
                payload = "trade_deny|$tradeId"
            ).sendAndWait(10000)
        }

        // Remove the pending trade
        pendingTrades.remove(tradeId)
        // Notify both players
        notifyTradeDenied(request.requesterId, targetId)

        return true
    }

    // Set up request listener for cross-server trade communications
    fun setupTradeRequestListener(currentServerId: String) {
        val jedis = JedisPool().resource

        val pubSub = object : JedisPubSub() {
            override fun onMessage(channel: String, message: String) {
                if (channel == "server:${currentServerId}:requests") {
                    val parts = message.split("|")
                    when (parts[0]) {
                        "trade_request" -> {
                            val dataKey = parts[1]
                            val tradeId = parts[2]
                            val requestData = Json.decodeFromString<TradeRequest>(jedis.get(dataKey))
                            // Store locally and notify player
                            pendingTrades[tradeId] = requestData
                            notifyPlayerOfTradeRequest(requestData.targetId, requestData.requesterId, tradeId)
                        }

                        "trade_accept" -> {
                            val dataKey = parts[1]
                            val tradeId = parts[2]

                            val request = pendingTrades[tradeId] ?: return

                            // Get requester's Pokémon
                            val requesterPokemon = getPlayerPokemonFromSlot(
                                request.requesterServer,
                                request.requesterId,
                                request.requesterPokemonSlot
                            )
                            // Get target's Pokémon from Redis
                            val targetPokemonData = jedis.get(dataKey)
                            val targetPokemon = deserializePokemon(targetPokemonData)

                            // Store requester's Pokémon in Redis for the target server
                            val requesterPokemonKey = "pokemontrade:response:$tradeId"
                            jedis.setex(requesterPokemonKey, 60, serializePokemon(requesterPokemon!!))

                            // Execute trade on this side
                            removePokemonFromPlayer(request.requesterId, request.requesterPokemonSlot)
                            addPokemonToPlayer(request.requesterId, targetPokemon)

                            // Respond with completion
                            val responseChannel = message.split("|").last()
                            jedis.publish(responseChannel, "COMPLETE|$requesterPokemonKey")

                            // Notify player
                            notifyTradeComplete(request.requesterId)

                            // Remove pending trade
                            pendingTrades.remove(tradeId)
                        }

                        "trade_deny" -> {
                            val tradeId = parts[1]
                            val request = pendingTrades[tradeId] ?: return

                            // Unlock Pokémon
                            unlockPokemon(request.requesterId, request.requesterPokemonSlot)

                            // Notify player
                            notifyTradeDenied(request.requesterId, request.targetId)

                            // Remove pending trade
                            pendingTrades.remove(tradeId)
                        }
                    }
                }
            }
        }

        // Start listening in a background thread
        Thread { jedis.subscribe(pubSub, "server:${currentServerId}:requests") }.start()
    }

    // In real implementation, we gonna get dara from player on a specified server via Velocity

    private fun getPlayerPokemonFromSlot(server: String, playerId: String, slot: Int): Pokemon? = Pokemon()
    private fun canTradePokemon(pokemon: Pokemon): Boolean = true

    private fun lockPokemon(server: String, playerId: String, slot: Int) {}
    private fun unlockPokemon(playerId: String, slot: Int) {}

    private fun scheduleTradeTimeout(tradeId: String) {}
    private fun executeTrade(requesterId: String, requesterSlot: Int, targetId: String, targetSlot: Int) {}

    private fun notifyPlayerOfTradeRequest(targetPlayerId: String, requesterId: String, tradeId: String) {}
    private fun notifyTradeComplete(playerId: String, targetId: String? = null) {}
    private fun notifyTradeFailed(playerId: String) {}
    private fun notifyTradeDenied(requesterId: String, targetId: String) {}

    private fun serializePokemon(pokemon: Pokemon): String = ""
    private fun deserializePokemon(data: String): Pokemon = Pokemon()

    private fun removePokemonFromPlayer(playerId: String, slot: Int) {}
    private fun addPokemonToPlayer(playerId: String, pokemon: Pokemon) {}
}